src/synthcity/plugins/core/models/tabnet.py

```diff
@@ -0,0 +1,831 @@
+            is_first = False
+
+        if n_glu_independent == 0:
+            # no independent layers
+            self.specifics = torch.nn.Identity()
+        else:
+            spec_input_dim = input_dim if is_first else output_dim
+            self.specifics = GLU_Block(
+                spec_input_dim, output_dim, first=is_first, **params  # type: ignore
+            )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = self.shared(x)
+        x = self.specifics(x)
+        return x
+
+
+class GLU_Block(torch.nn.Module):
+    """
+    Independent GLU block, specific to each step
+    """
+
+    def __init__(
+        self,
+        input_dim: int,
+        output_dim: int,
+        *,
+        n_glu: int = 2,
+        first: bool = False,
+        shared_layers: Optional[torch.nn.ModuleList] = None,
+        virtual_batch_size: int = 128,
+        momentum: float = 0.02,
+    ) -> None:
+        super(GLU_Block, self).__init__()
+        self.first = first
+        self.shared_layers = shared_layers
+        self.n_glu = n_glu
+        self.glu_layers = torch.nn.ModuleList()
+
+        params = {"virtual_batch_size": virtual_batch_size, "momentum": momentum}
+
+        fc = shared_layers[0] if shared_layers else None
+        self.glu_layers.append(GLU_Layer(input_dim, output_dim, fc=fc, **params))  # type: ignore
+        for glu_id in range(1, self.n_glu):
+            fc = shared_layers[glu_id] if shared_layers else None
+            self.glu_layers.append(GLU_Layer(output_dim, output_dim, fc=fc, **params))  # type: ignore
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        scale = torch.sqrt(torch.FloatTensor([0.5]).to(x.device))
+        if self.first:  # the first layer of the block has no scale multiplication
+            x = self.glu_layers[0](x)
+            layers_left = range(1, self.n_glu)
+        else:
+            layers_left = range(self.n_glu)
+
+        for glu_id in layers_left:
+            x = torch.add(x, self.glu_layers[glu_id](x))
+            x = x * scale
+        return x
+
+
+class GLU_Layer(torch.nn.Module):
+    def __init__(
+        self,
+        input_dim: int,
+        output_dim: int,
+        *,
+        fc: Optional[Linear] = None,
+        virtual_batch_size: int = 128,
+        momentum: float = 0.02,
+    ) -> None:
+        super(GLU_Layer, self).__init__()
+
+        self.output_dim = output_dim
+        if fc:
+            self.fc = fc
+        else:
+            self.fc = Linear(input_dim, 2 * output_dim, bias=False)
+        initialize_glu(self.fc, input_dim, 2 * output_dim)
+
+        self.bn = GBN(
+            2 * output_dim, virtual_batch_size=virtual_batch_size, momentum=momentum
+        )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = self.fc(x)
+        x = self.bn(x)
+        out = torch.mul(x[:, : self.output_dim], torch.sigmoid(x[:, self.output_dim :]))
+        return out
+
+
+# class TabNetDecoder(torch.nn.Module):
```

Comment:
Please remove these comments if the code is not used

Reply:
Done.